lib/commanded/event/handler.ex

```diff
@@ -533,6 +534,20 @@ defmodule Commanded.Event.Handler do
       @before_compile unquote(__MODULE__)
       @behaviour Handler
 
+      @on_error (case Keyword.get(unquote(using_opts), :on_error, :stop) do
+                   :stop ->
+                     &ErrorHandler.stop_on_error/3
+
+                   :backoff ->
+                     &ErrorHandler.backoff/3
+
+                   :grumpy_old_bastard ->
+                     &ErrorHandler.grump/3
+
+                   error_handler when is_function(error_handler, 3) ->
+                     error_handler
+                 end)
```

Comment:
I rather have `retry_policy` + `error` callback that this to be honest, otherwise, `error` callback sounds good enough

I am not sure, I don't like the API somewhat like this

Reply:
I'm not super sure what you are saying here, but I agree that something is confusing here. Having an `error/3` callback... but also an `on_error` mechanism is redundant.

We need to keep the `error/3` callback...

Reply:
We need to decide if the retry mechanism will trigger before it reaches the error callback.

Can we do that first?

Reply:
`error/3` is meant to be the place where you customize how you deal with errors in an event handler. That is an excellent option for customizing the behaviour of 1 event handler.

If we had no restrictions, I think making exponential backoff the default would be the best option. We can't do that because that would be a surprising and confusing breaking change.

I don't think we can separate retrying from the error callback, they are linked and I think that is appropriate. It would be nice if you could specify either the configuration option OR a custom callback, and get an error if you try to do both. I haven't found a good way of doing that yet.

We could walk back a step and only provide implementations functions for backoff and other strategies that you could call from `error/3`:

```elixir
@impl true
def error(_error, _event, _context) do
  backoff(context, max_retries: 42)
end
```
It's not as nice as the configuration you suggested, but much less tangled.

Finally, I wanted to make this a change to the default so that all of the handlers in a project would have this behaviour without having to remember to opt-in to it. What if we made a global or Commanded application level configuration that would apply to all of the handlers which would apply unless you implement `error/3`. This would signal your desire to override the configured behaviour.

Something like:
```elixir
# config/config.exs

config :commanded,
  default_error_handler: :backoff

# or

config MyApp.CommandedApp,
  default_error_handler: :backoff 

```

Reply:
> We can't do that because that would be a surprising and confusing breaking change.

Could we do it in a major version bump, though? I agree that would be a better out-of-the-box experience because "event handler error takes down entire app" is one of the biggest complaints I've encountered from folks who are new at working with OTP and don't yet understand how to deal with it on a supervision tree level.

Reply:
N.B. I wouldn't hold up *this* PR for that major version change discussion; that was more of an aside.

Reply:
I think if we do it this way, we can swap the default easily in a 2.0.